### PR TITLE
JDK-8301354: Modernize utilities/breakpoint.hpp

### PR DESCRIPTION
**Pull request by jcking**
Cleanup `utilities/breakpoint.hpp` by utilizing compiler intrinsics when available, or normal signals for AIX/XLC.

---------
### Progress
- [ ] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8301354](https://bugs.openjdk.org/browse/JDK-8301354): Modernize utilities/breakpoint.hpp

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/12296/head:pull/12296` \
`$ git checkout pull/12296`

Update a local copy of the PR: \
`$ git checkout pull/12296` \
`$ git pull https://git.openjdk.org/jdk pull/12296/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 12296`

View PR using the GUI difftool: \
`$ git pr show -t 12296`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/12296.diff">https://git.openjdk.org/jdk/pull/12296.diff</a>

</details>

### CODE REVIEW

**Comment by jcking** [comment]
Converting back to draft while I investigate some things.

File: src/hotspot/share/utilities/breakpoint.hpp [hunk 1]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 2017, 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 2023, Google and/or its affiliates. All rights reserved.
```
Comment (dholmes-ora) [inline src/hotspot/share/utilities/breakpoint.hpp:2]: Please restore the Oracle copyright with second year updated to 2023. Thanks.

File: src/hotspot/share/utilities/breakpoint.hpp [hunk 2]
```diff
@@ -25,11 +25,52 @@
...
+
+#define BREAKPOINT __debugbreak()
+
+#elif defined(TARGET_COMPILER_xlc)
+
+#if defined(SIGTRAP)
+
+#define BREAKPOINT ::raise(SIGTRAP)
+
+#elif defined(SIGINT)
```
Comment (dholmes-ora) [inline src/hotspot/share/utilities/breakpoint.hpp:60]: The VM already uses SIGINT to shutdown the VM - raising it here doesn't seem right.

**Review by dholmes-ora — COMMENTED** [review]
So you are turning something currently OS-specific into something toolchain specific, and I'm not sure that is really correct, or even necessary.

**Comment by tstuefe** [comment]
Hi Justin,

please note that at least on PPC platforms we use SIGTRAP for implicit null checks; so gcc debug trap won't work as expected.

Also, IIUC this changes os::breakpoint() on GCC platforms and AIX from being a manual opt-in to an unconditional trap right? While this is more in sync with what we do on Windows, this behavior has been in there a long time and may confuse people.

Finally, there is this worrying info: https://gcc.gnu.org/bugzilla/show_bug.cgi?id=84595 where it seems the GCC optimizes everything after a trap away? Can that happen with debugtrap too?

..Thomas

**Comment by jcking** [comment]
I put this back into draft after I realized that I forgot only Clang has `__builtin_debugtrap`, making this non-ideal. Ideally GCC would have that as well.

In an ideal world, breakpoint should probably only be present in debug builds and compiled out in release builds. Additionally it should probably be consistent across platforms. But I think that ship has sailed. If GCC ever gets `__builtin_debugtrap` than this would be useful, IIF we make them used during debug builds only and fallback to `breakpoint()` calls for everything else.

**Pull request closed without merging**